client/client.go

```diff
@@ -383,25 +393,45 @@ func (r *NotaryRepository) Publish() error {
 		return err
 	}
 
+	// these are the tuf files we will need to update, serialized as JSON before
+	// we send anything to remote
+	update := make(map[string][]byte)
+
 	// check if our root file is nearing expiry. Resign if it is.
-	if nearExpiry(r.tufRepo.Root) || r.tufRepo.Root.Dirty {
-		if err != nil {
-			return err
-		}
+	if nearExpiry(r.tufRepo.Root) || r.tufRepo.Root.Dirty || updateRoot {
 		root, err = r.tufRepo.SignRoot(data.DefaultExpires("root"))
+
+		rootJSON, err := json.Marshal(root)
 		if err != nil {
 			return err
 		}
-		updateRoot = true
+		update[data.CanonicalRootRole] = rootJSON
 	}
-	// we will always resign targets and snapshots
-	targets, err := r.tufRepo.SignTargets("targets", data.DefaultExpires("targets"))
+
+	// we will always resign targets, and snapshots if we have the snapshots key
+	targetsJSON, err := serializeCanonicalRole(r.tufRepo, data.CanonicalTargetsRole)
 	if err != nil {
 		return err
 	}
-	snapshot, err := r.tufRepo.SignSnapshot(data.DefaultExpires("snapshot"))
-	if err != nil {
-		return err
+	update[data.CanonicalTargetsRole] = targetsJSON
+
+	// do not update the snapshot role if we do not have the snapshot key or
+	// any snapshot data.  There might not be any snapshot data the repo was
+	// initialized with the snapshot signing role delegated to the server.
+	// The repo might have snapshot data, because it was requested from
+	// the server by listing, but not have the snapshot key, so signing will
+	// fail.
+	if r.tufRepo.Snapshot != nil {
+		snapshotJSON, err := serializeCanonicalRole(
+			r.tufRepo, data.CanonicalSnapshotRole)
+		if err == nil { // we have the key - snapshot signed, let's update it
+			update[data.CanonicalSnapshotRole] = snapshotJSON
+		} else if _, ok := err.(signed.ErrNoKeys); ok {
+			logrus.Debugf("Client does not have the key to sign snapshot. " +
```

Comment:
Unless we update the TUF files to add extra state about who has the key, all we can do is know that we do not have the key, not that the server is supposed to have it.  We can't tell whether we just lost it or whether we're not supposed to have it.